std/algorithm/iteration.d

```diff
@@ -4035,16 +4038,97 @@ if (isInputRange!R && !isInfinite!R && is(typeof(seed = seed + r.front)))
 }
 
 // Pairwise summation http://en.wikipedia.org/wiki/Pairwise_summation
-private auto sumPairwise(Result, R)(R r)
+private auto sumPairwise(F, R)(R data)
+if (isInputRange!R && !isInfinite!R)
 {
-    static assert (isFloatingPoint!Result);
-    switch (r.length)
+    import core.bitop : bsf;
+    // Works for r with at least length < 2^^(64 + log2(16)), in keeping with the use of size_t
+    // elsewhere in std.algorithm and std.range on 64 bit platforms. The 16 in log2(16) comes
+    // from the manual unrolling in sumPairWise16
+    F[64] store = void;
+    size_t idx = 0;
+
+    auto collapseStore(T)(T k)
```

Comment:
@andralex On x86 or on x64 as well?
On 24 Mar 2016 8:04 pm, "Andrei Alexandrescu" <notifications@github.com>
wrote:

> In std/algorithm/iteration.d
> <https://github.com/D-Programming-Language/phobos/pull/4069#discussion_r57379092>
> :
>
> >  {
> > -    static assert (isFloatingPoint!Result);
> > -    switch (r.length)
> > +    import core.bitop : bsf;
> > +    // Works for r with at least length < 2^^(64 + log2(16)), in keeping with the use of size_t
> > +    // elsewhere in std.algorithm and std.range on 64 bit platforms. The 16 in log2(16) comes
> > +    // from the manual unrolling in sumPairWise16
> > +    F[64] store = void;
> > +    size_t idx = 0;
> > +
> > +    auto collapseStore(T)(T k)
>
> Iota with longs should probably not offer length.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/D-Programming-Language/phobos/pull/4069/files/ef9f8338fe123c1514a61c5b4b50fbfb072632e4#r57379092>
>